### Capstone Project - Final 

## Segmenting and Clustering Neighbourhoods of the City of Toronto

In [1]:
import pandas as pd
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
import matplotlib as mpl
%matplotlib inline
import seaborn as sns


import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

!pip install pandas==1.0.3

import pandas as pd



#!pip install geocoder
#import geocoder # import geocoder

# initialize your variable to None
#lat_lng_coords = None

# loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(new_data['Postal Code']))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Libraries imported.


In [2]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


## PART 1 - Building the Neighbourhood DataFrame

#### WEB SCRAPING

In [3]:
!pip install lxml

In [4]:
raw_data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M", header=0)
raw_data = raw_data[0]
raw_data.head(10)

Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

#### Filtering only valid Boroughs - Removing Not Assigned Boroughs

In [5]:
new_data = raw_data[raw_data['Borough']!='Not assigned']
new_data.shape

(103, 3)

In [6]:
new_data.reset_index(drop=True,inplace=True)

In [7]:
new_data.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [8]:
new_data[new_data['Neighbourhood']=='Not Assigned']

# No Neighbourhoods Exist with NOT ASSIGNED if they have a borough

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [9]:
new_data.shape

(103, 3)

## PART 2 - Including the Latitude and Longitude

#### Getting Latitude and Longitude of Neighbourhoods - Google Geocoder is not complying

In [10]:
latlong = pd.read_csv('Geospatial_Coordinates.csv') 
latlong.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Joining Neighbourhoods with their Latitudes and Longitudes

In [11]:
neigh_with_latlong = new_data.merge(right=latlong,on='Postal Code')
neigh_with_latlong.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

## Part 3 - Exploring and Segmenting Neighbourhoods and Venues

In [12]:
len(neigh_with_latlong['Borough'].unique())

10

#### > We have 10 unique boroughs

In [13]:
neigh_with_latlong['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

In [14]:
len(neigh_with_latlong['Neighbourhood'])

103

In [15]:
neigh_with_latlong[neigh_with_latlong[['Neighbourhood']].duplicated()]

Postal Code     Borough Neighbourhood   Latitude  Longitude
13         M3C  North York     Don Mills  43.725900 -79.340923
46         M3L  North York     Downsview  43.739015 -79.506944
53         M3M  North York     Downsview  43.728496 -79.495697
60         M3N  North York     Downsview  43.761631 -79.520999

#### > We have 99 unique neighbourhoods and 103 total neighbourhoods with different postal codes

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [16]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### lets visualize ALL Boroughs and Neighbourhoods around toronto

In [17]:
# create map of New York using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neigh_with_latlong['Latitude'], neigh_with_latlong['Longitude'], neigh_with_latlong['Borough'], neigh_with_latlong['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

#### lets visualize ONLY Boroughs around Downtown toronto and nearby Boroughs

In [18]:
toronto_df = neigh_with_latlong[neigh_with_latlong['Borough'].str.contains('Toronto')]
toronto_df.head(3)

Postal Code           Borough                                Neighbourhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
2  43.654260 -79.360636  
4  43.662301 -79.389494  
9  43.657162 -79.378937

In [19]:
# create map of New York using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

### Exploring Venues around Downtown Toronto and Nearby Boroughs

In [20]:
# @hidden_cell

CLIENT_ID = '0IRSU1PXCTADSJ0GAWTDUVNWMXS352WY4JU4XJN4XBLW4BV1' # your Foursquare ID
CLIENT_SECRET = 'TCNAWWPCZ22X34PMOHGEA2O0S1BILYL3BS3TQRD1Y5YQDAUG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT= 100
radius = 500

Defining a function that retrieves venues for each neighbourhood and creates a dataframe

In [21]:
def getNearbyVenues(postal, borough, names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for postal, borough, name, lat, lng in zip(postal, borough, names, latitudes, longitudes):
        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            postal,
            borough,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = [
                  'Postal Code',
                  'Borough',
                  'Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


In [22]:
toronto_venues = getNearbyVenues(  postal = toronto_df['Postal Code'],
                                   borough = toronto_df['Borough'],
                                   names = toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

#### The Compelete DataFrame containing Downtown Toronto Boroughs, Nearby Boroughs and their VENUES

In [23]:
print(toronto_venues.shape)
toronto_venues.head(5)

(1638, 9)


Postal Code           Borough              Neighbourhood  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront   
1         M5A  Downtown Toronto  Regent Park, Harbourfront   
2         M5A  Downtown Toronto  Regent Park, Harbourfront   
3         M5A  Downtown Toronto  Regent Park, Harbourfront   
4         M5A  Downtown Toronto  Regent Park, Harbourfront   

   Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                43.65426               -79.360636        Roselle Desserts   
1                43.65426               -79.360636           Tandem Coffee   
2                43.65426               -79.360636  Cooper Koo Family YMCA   
3                43.65426               -79.360636     Body Blitz Spa East   
4                43.65426               -79.360636          Impact Kitchen   

   Venue Latitude  Venue Longitude       Venue Category  
0       43.653447       -79.362017               Bakery  
1       43.653559       -79.361809          Coffee Shop  
2       43.653249       -79.358008  Distribution Center  
3       43.654735       -79.359874                  Spa  
4       43.656369       -79.356980           Restaurant

In [24]:
len(toronto_venues[toronto_venues['Borough']=='Downtown Toronto']['Neighbourhood'].unique())

19

In [25]:
toronto_venues.groupby('Neighbourhood').count()

Postal Code  Borough  \
Neighbourhood                                                              
Berczy Park                                                  57       57   
Brockton, Parkdale Village, Exhibition Place                 24       24   
Business reply mail Processing Centre, South Ce...           17       17   
CN Tower, King and Spadina, Railway Lands, Harb...           16       16   
Central Bay Street                                           65       65   
Christie                                                     17       17   
Church and Wellesley                                         76       76   
Commerce Court, Victoria Hotel                              100      100   
Davisville                                                   34       34   
Davisville North                                              8        8   
Dufferin, Dovercourt Village                                 17       17   
First Canadian Place, Underground city                      100      100   
Forest Hill North & West, Forest Hill Road Park               4        4   
Garden District, Ryerson                                    100      100   
Harbourfront East, Union Station, Toronto Islands           100      100   
High Park, The Junction South                                24       24   
India Bazaar, The Beaches West                               19       19   
Kensington Market, Chinatown, Grange Park                    66       66   
Lawrence Park                                                 3        3   
Little Portugal, Trinity                                     45       45   
Moore Park, Summerhill East                                   2        2   
North Toronto West, Lawrence Park                            19       19   
Parkdale, Roncesvalles                                       14       14   
Queen's Park, Ontario Provincial Government                  36       36   
Regent Park, Harbourfront                                    47       47   
Richmond, Adelaide, King                                    100      100   
Rosedale                                                      4        4   
Roselawn                                                      1        1   
Runnymede, Swansea                                           36       36   
St. James Town                                               87       87   
St. James Town, Cabbagetown                                  45       45   
Stn A PO Boxes                                               97       97   
Studio District                                              40       40   
Summerhill West, Rathnelly, South Hill, Forest ...           16       16   
The Annex, North Midtown, Yorkville                          20       20   
The Beaches                                                   5        5   
The Danforth West, Riverdale                                 42       42   
Toronto Dominion Centre, Design Exchange                    100      100   
University of Toronto, Harbord                               35       35   

                                                    Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                         8   

#### Lets Analyze Each Neighbourhood within Downtown Toronto Borough and other nearby Boroughs (Containing TORONTO)

In [26]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1638, 236)


Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Stop  Butcher  Café  \
0             0              0         0         0        0     0   
1             0              0         0         0        0     0   
2             0              0         0         0        0     0   
3             0              0         0         0        0

#### Lets Group by Neighbourhood - Taking the mean of the frequency of occurrence of each category - Reduce the DataFrame Size

In [27]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()

print(toronto_grouped.shape)
toronto_grouped.head()

(39, 236)


Neighbourhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0   0.0000              0.0000        0.0000           0.000           0.0000   
1   0.0000              0.0000        0.0000           0.000           0.0000   
2   0.0000              0.0000        0.0000           0.000           0.0000   
3   0.0625              0.0625        0.0625           0.125           0.0625   
4   0.0000              0.0000        0.0000           0.000           0.0000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0             0.000                  0.0           0.0       0.0     0.017544   
1             0.000                  0.0           0.0       0.0     0.000000   
2             0.000                  0.0           0.0       0.0     0.000000   
3             0.125                  0.0           0.0       0.0     0.000000   
4             0.000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.015385                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0       0.000000   0.017544         0.0    0.017544  0.035088   0.0  0.000000   
1       0.000000   0.000000         0.0    0.000000  0.041667   0.0  0.041667   
2       0.058824   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
3       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.062500   
4       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.015385   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.017544  0.017544              0.0  0.035088   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Board Shop  Boat or Ferry  \
0         0.0                 0.0  0.017544         0.0         0.0000   
1         0.0                 0.0  0.000000         0.0         0.0000   
2         0.0                 0.0  0.000000         0.0         0.0000   
3         0.0                 0.0  0.000000         0.0         0.0625   
4         0.0                 0.0  0.000000         0.0         0.0000   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  \
0        0.0    0.0000                   0.0        0.017544  0.000000   
1        0.0    0.0000                   0.0        0.083333  0.000000   
2        0.0    0.0000                   0.0        0.000000  0.058824   
3        0.0    0.0625                   0.0        0.000000  0.000000   
4        0.0    0.0000                   0.0        0.000000  0.000000   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Bus Stop  \
0         0.000000       0.0      0.000000       0.000000       0.0       0.0   
1         0.000000       0.0      0.000000       0.041667       0.0       0.0   
2         0.000000       0.0      0.000000       0.05882

#### Let's print each neighborhood along with the top 5 most common venues

In [28]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1            Beer Bar  0.04
2        Cocktail Bar  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2     Coffee Shop  0.08
3             Gym  0.04
4   Grocery Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0    Light Rail Station  0.12
1            Comic Shop  0.06
2         Auto Workshop  0.06
3  Fast Food Restaurant  0.06
4            Smoke Shop  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1  Airport Terminal  0.12
2          Boutique  0.06
3   Harbor / Marina  0.06
4     Boat or Ferry  0.06


----Central Bay Street----
                venue  freq
0       

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

# This function will take one row at a time, arranges the values in descending order, and then takes its index values, which are the categories names of those values but only the highest 10
# Returns the Top 10 categories and we put those names in different columns for each neighbourhood

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery        Farmers Market              Beer Bar   
1        Breakfast Spot           Coffee Shop           Yoga Studio   
2           Pizza Place            Skate Park               Brewery   
3      Airport Terminal                   Bar       Harbor / Marina   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0            Restaurant          Cocktail Bar            Cheese Shop   
1         Grocery Store             Pet Store  Performing Arts Venue   
2                   Spa         Burrito Place         Farmers Market   
3      Sculpture Garden         Boat or Ferry            Coffee Shop   
4   Japanese Restaurant           Salad Place           Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café    Seafood Restaurant                  Hotel  
1             Nightclub    Italian Restaurant           Intersection  
2  Fast Food Restaurant            Restaurant          Auto Workshop  
3   Rental Car Location              Boutique                  Plane  
4      Department Store       Bubble Tea Shop            Yoga Studio

## Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [31]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0,n_init=10).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:30] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 3, 0,
       0, 0, 0, 0, 4, 1, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [32]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.merge(right=neighborhoods_venues_sorted, on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               2      Asian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Café                   Pub   
1           Yoga Studio   Fried Chicken Joint            Restaurant   
2           Coffee Shop                  Café   Japanese Restaurant   
3                  Café        Cosmetics Shop            Restaurant   
4                   Pub     Health Food Store                 Trail   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Park               Theater        Breakfast Spot   
1  Portuguese Restaurant    Persian Restaurant                  Park   
2        Bubble Tea Shop        Cosmetics Shop               Theater   
3           Cocktail Bar   American Restaurant                 Diner   
4           Neighborhood           Yoga Studio        Discount Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant     Electronics Store             Shoe Store  
1           Music Venue    Mexican Restaurant    Japanese Restaurant  
2        Lingerie Store      Ramen Restaurant     Italian Restaurant  
3    Seafood Restaurant              Beer Bar                   Park  
4      Department Store          Dessert Shop                  Diner

Finally, let's visualize the resulting clusters

In [33]:
# create map
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude


map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, We can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0              Pharmacy   
10  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0                 Bakery                    Café                   Pub   
1            Yoga Studio     Fried Chicken Joint            Restaurant   
2            Coffee Shop                    Café   Japanese Restaurant   
3                   Café          Cosmetics Shop            Restaurant   
5                 Bakery          Farmers Market              Beer Bar   
6                   Café          Sandwich Place    Italian Restaurant   
7                   Café                    Park           Candy Store   
8         Clothing Store                    Café                 Hotel   
9                 Bakery  Furniture / Home Store                  Bank   
10              Aquarium                   Hotel                  Café   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                        Park               Theater        Breakfast Spot   
1       Portuguese Restaurant    Persian Restaurant                  Park   
2             Bubble Tea Shop        Cosmetics Shop               Theater   
3                Cocktail Bar   American Restaurant                 Diner   
5                  Restaurant          Cocktail Bar           Cheese Shop   
6         Japanese Restaurant           Salad Place          Burger Joint   
7                  Restaurant    Athletics & Sports                 Diner   
8                         Gym            Restaurant                   Bar   
9   Middle Eastern Restaurant           Supermarket                  Café   
10                 Restaurant   Fried Chicken Joint        Scenic Lookout   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Restaurant     Electronics Store             Shoe Store  
1            Music Venue    Mexican Restaurant    Japanese Restaurant  
2         Lingerie Store      Ramen Restaurant     Italian Restaurant  
3     Seafood Restaurant              Beer Bar                   Park  
5                   Café    Seafood Restaurant                  Hotel  
6       Department Store       Bubble Tea Shop            Yoga Studio  
7     Italian Restaurant            Baby Store              Nightclub  
8        Thai Restaurant            Steakhouse            Pizza Place  
9               Bus Stop           Pizza Place                    Bar  
10    Italian Restaurant               Brewery            Music Venue

In [35]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,15:].T.iloc[:,0].sort_values(ascending=False).head(10)

Coffee Shop           0.080808
Café                  0.054794
Park                  0.038702
Restaurant            0.032048
Pizza Place           0.025018
Italian Restaurant    0.024343
Sandwich Place        0.020066
Bar                   0.019658
Bakery                0.019181
Pub                   0.016043
Name: 0, dtype: float64

#### Cluster 2

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Yoga Studio          Dance Studio     Electronics Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19      Doner Restaurant               Dog Run    Distribution Center

In [37]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,15:].T.iloc[:,1].sort_values(ascending=False).head(10)

Garden                  1.0
Yoga Studio             0.0
Food Court              0.0
Flea Market             0.0
Fish Market             0.0
Fish & Chips Shop       0.0
Filipino Restaurant     0.0
Fast Food Restaurant    0.0
Farmers Market          0.0
Falafel Restaurant      0.0
Name: 1, dtype: float64

#### Cluster 3

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  East Toronto               2      Asian Restaurant                   Pub   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4     Health Food Store                 Trail          Neighborhood   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4           Yoga Studio        Discount Store      Department Store   

  9th Most Common Venue 10th Most Common Venue  
4          Dessert Shop                  Diner

In [39]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,15:].T.iloc[:,2].sort_values(ascending=False).head(10)

Asian Restaurant        0.2
Trail                   0.2
Pub                     0.2
Neighborhood            0.2
Health Food Store       0.2
Grocery Store           0.0
Greek Restaurant        0.0
Fast Food Restaurant    0.0
Farmers Market          0.0
Falafel Restaurant      0.0
Name: 2, dtype: float64

#### Cluster 4

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               3                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29           Summer Camp           Yoga Studio          Dance Studio   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
29     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29            Donut Shop      Doner Restaurant                Dog Run

In [41]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,15:].T.iloc[:,3].sort_values(ascending=False).head(10)

Trail                   0.5
Summer Camp             0.5
Yoga Studio             0.0
Donut Shop              0.0
Fish Market             0.0
Fish & Chips Shop       0.0
Filipino Restaurant     0.0
Fast Food Restaurant    0.0
Farmers Market          0.0
Falafel Restaurant      0.0
Name: 3, dtype: float64

#### Cluster 5

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
21   Central Toronto               4                  Park   
33  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Jewelry Store                 Trail      Sushi Restaurant   
33            Playground                 Trail          Dance Studio   

   5th Most Common Venue        6th Most Common Venue  \
21           Yoga Studio                Deli / Bodega   
33     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
21  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
33          Dumpling Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
21       Doner Restaurant  
33                Dog Run

In [43]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,15:].T.iloc[:,4].sort_values(ascending=False).head(10)

Park                    0.375
Trail                   0.250
Playground              0.125
Jewelry Store           0.125
Sushi Restaurant        0.125
Yoga Studio             0.000
Donut Shop              0.000
Filipino Restaurant     0.000
Fast Food Restaurant    0.000
Farmers Market          0.000
Name: 4, dtype: float64

## Conclusion 
### Based on the clusters we can see the following:
1. Cluster 1: Downtown style cafe shops and resturants
2. Cluster 2: Only Gardens
3. Cluster 3: Out of downtown Areas Asian style restaurants and PUBs
4. Cluster 4: Trails and Summer Camps Areas
5. Cluster 5: Parks and Trails